### PR TITLE
ocs-to-ocs: check if ticket date is greater than current date

### PR DESCRIPTION
**Pull request by subhamkrai**
adding check to verify if if ticket date is greater than current date.

Signed-off-by: subhamkrai <srai@redhat.com>

### CODE REVIEW

File: services/provider/server/server.go [hunk 1]
```diff
@@ -412,6 +418,18 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
...
+	var ticketDate TicketDate
+	err = json.Unmarshal(message, &ticketDate)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal message. %v", err)
+	}
+
+	if ticketDate.ExpirationDate < time.Now().Unix() {
+		return fmt.Errorf("%s id %d date is lesser than the current date", ticketDate.ID, ticketDate.ExpirationDate)
+	}
+	klog.Infof("onboarding ticket %s has been verified successfully", ticketDate.ID)
```
Comment (nb-ohad) [inline services/provider/server/server.go:431]: Can we move this after the signature check

File: services/provider/server/server.go [hunk 2]
```diff
@@ -412,6 +418,18 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
...
 	}
+
+	var ticketDate TicketDate
+	err = json.Unmarshal(message, &ticketDate)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal message. %v", err)
+	}
+
+	if ticketDate.ExpirationDate < time.Now().Unix() {
+		return fmt.Errorf("%s id %d date is lesser than the current date", ticketDate.ID, ticketDate.ExpirationDate)
```
Comment (nb-ohad) [inline services/provider/server/server.go:429]: Message can use a rephrase, maybe:
`Onboarding ticket (<id>) is expired `

File: services/provider/server/server.go [hunk 3]
```diff
@@ -412,6 +418,18 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
...
+	}
+	klog.Infof("onboarding ticket %s has been verified successfully", ticketDate.ID)
+
 	signature, err := base64.StdEncoding.DecodeString(ticketArr[1])
 	if err != nil {
 		return fmt.Errorf("failed to decode signature. %v", err)
```
Comment (nb-ohad) [inline services/provider/server/server.go:435]: Incorporate id into message, maybe `failed to decode onboarding ticket (<id>) signature: %v`

File: services/provider/server/server.go [hunk 4]
```diff
@@ -412,10 +418,22 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
...
 	signature, err := base64.StdEncoding.DecodeString(ticketArr[1])
 	if err != nil {
-		return fmt.Errorf("failed to decode signature. %v", err)
+		return fmt.Errorf("failed to decode onboarding ticket %s signature: %v", ticketDate.ID, err)
+	}
+
+	if ticketDate.ExpirationDate < time.Now().Unix() {
+		return fmt.Errorf(" Onboarding ticket %s is expired ", ticketDate.ID)
 	}
+	klog.Infof("onboarding ticket %s has been verified successfully", ticketDate.ID)
```
Comment (nb-ohad) [inline services/provider/server/server.go:436]: This need to be the last thing on the function, and there is another variation of it already there `Successfully verified onboarding ticket`, but with no ID

**Review by nb-ohad — APPROVED** [review]

File: services/provider/server/server.go [hunk 5]
```diff
@@ -412,9 +418,20 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
 	if err != nil {
 		return fmt.Errorf("failed to decode payload. %v", err)
 	}
+
+	var ticketDate TicketDate
+	err = json.Unmarshal(message, &ticketDate)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal message. %v", err)
```
Comment (nb-ohad) [inline services/provider/server/server.go:425]:
```suggestion
		return fmt.Errorf("failed to unmarshal onboarding ticket message. %v", err)
```

File: services/provider/server/server.go [hunk 6]
```diff
@@ -412,9 +418,20 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
 	if err != nil {
 		return fmt.Errorf("failed to decode payload. %v", err)
```
Comment (nb-ohad) [inline services/provider/server/server.go:419]:
```suggestion
		return fmt.Errorf("failed to decode onboarding ticket: %v", err)
```

File: services/provider/server/server.go [hunk 7]
```diff
@@ -423,7 +436,11 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
 		return fmt.Errorf("failed to verify onboarding ticket signature. %v", err)
 	}
 
-	klog.Info("Successfully verified onboarding ticket")
+	if ticketData.ExpirationDate < time.Now().Unix() {
+		return fmt.Errorf(" Onboarding ticket %s is expired ", ticketData.ID)
```
Comment (nb-ohad) [inline services/provider/server/server.go:440]:
```suggestion
		return fmt.Errorf("Onboarding ticket %s is expired", ticketData.ID)
```

**Review by jarrpa — CHANGES_REQUESTED** [review]
Requesting a few language changes, and also an update to the hashing algorithm to match with the latest `ticketgen.sh`.

File: services/provider/server/server.go [hunk 8]
```diff
@@ -412,9 +418,20 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
 	if err != nil {
 		return fmt.Errorf("failed to decode payload. %v", err)
 	}
+
+	var ticketDate TicketDate
```
[the remark this answers is not in the document]
Comment (subhamkrai) [reply services/provider/server/server.go]: @jarrpa we `ticket` is already used changed it to `ticketData`

**Review by nb-ohad — APPROVED** [review]

File: services/provider/server/server.go [hunk 9]
```diff
@@ -410,20 +416,31 @@ func validateTicket(ticket string, pubKey *rsa.PublicKey) error {
...
+
 	signature, err := base64.StdEncoding.DecodeString(ticketArr[1])
 	if err != nil {
-		return fmt.Errorf("failed to decode signature. %v", err)
+		return fmt.Errorf("failed to decode onboarding ticket %s signature: %v", ticketData.ID, err)
 	}
 
-	hash := md5.Sum(message)
+	hash := sha256.Sum256(message)
 	err = rsa.VerifyPKCS1v15(pubKey, crypto.MD5, hash[:], signature)
```
Comment (jarrpa) [inline services/provider/server/server.go:434]: As seen here: https://github.com/jarrpa/ocs-operator/blob/ticketgen/hack/ticketgen/ticketcheck.go#L57
```suggestion
	err = rsa.VerifyPKCS1v15(pubKey, crypto.SHA256, hash[:], signature)
```

**Review by jarrpa — CHANGES_REQUESTED** [review]
/lgtm

**Review by jarrpa — APPROVED** [review]
/lgtm

**Pull request merged**